src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp

```diff
@@ -0,0 +1,945 @@
+    Common_.halt_if_singular=0;
+    //allocate gpu data
+    free(mia_); 
+    free(mja_);
+    mia_ = NULL; 
+    mja_ = NULL;
+    cudaFree(devx_);
+    cudaFree(devr_);
+    devx_ = NULL;
+    devr_ = NULL;
+    checkCudaErrors(cudaMalloc(&devx_, n_ * sizeof(double)));
+    checkCudaErrors(cudaMalloc(&devr_, n_ * sizeof(double)));
+    this->newKLUfactorization();
+  }
+
+  //helper private function needed for format conversion
+
+  int hiopLinSolverIndefSparseCUSOLVER::createM(const int n, 
+                                                const int nnzL, 
+                                                const int* Lp, 
+                                                const int* Li,
+                                                const int nnzU, 
+                                                const int* Up, 
+                                                const int* Ui){
+
+    int row;
+    for(int i = 0;i<n;++i) {
+      //go through EACH COLUMN OF L first
+      for(int j = Lp[i]; j<Lp[i+1]; ++j) {
+        row = Li[j];
+        //BUT dont count dia_gonal twice, important
+        if(row != i) {
+          mia_[row + 1]++;
+        }
+      }
+      //then each column of U
+      for(int j = Up[i];j<Up[i+1];++j) {
+        row = Ui[j];
+        mia_[row + 1]++;
+      }
+    }
+    //then organize mia_;
+    mia_[0] = 0;
+    for(int i=1;i<n+1;i++) {
+      mia_[i] += mia_[i-1];
+    } 
+
+    int* Mshifts = (int*) calloc (n,sizeof(int));
+    for(int i = 0;i<n;++i) {
+
+      //go through EACH COLUMN OF L first
+      for(int j = Lp[i];j<Lp[i+1];++j) {
+        row = Li[j];
+        if(row != i) {
+          //place (row, i) where it belongs!
+
+          mja_[mia_[row] + Mshifts[row]] = i;
+          Mshifts[row]++;
+        }
+      }
+      //each column of U next
+      for(int j = Up[i];j<Up[i+1];++j) {
+        row = Ui[j];
+        mja_[mia_[row] + Mshifts[row]] = i;
+        Mshifts[row]++;
+      }
+    }
+    free(Mshifts);
+    return 0;
+  }
+
+
+  // call if both the matrix and the nnz structure changed or if convergence is poor while using refactorization.
+  int hiopLinSolverIndefSparseCUSOLVER::newKLUfactorization()
+  {
+    klu_free_symbolic(&Symbolic_, &Common_) ;
+    klu_free_numeric(&Numeric_, &Common_) ;
+    Symbolic_ = klu_analyze(n_, kRowPtr_, jCol_, &Common_) ;
+    if(Symbolic_ == NULL) {
+      nlp_->log->printf(hovError, "symbolic NULL\n");
+      return -1;
+    }
+
+    Numeric_ = klu_factor(kRowPtr_, jCol_, kVal_, Symbolic_, &Common_);
+    if(Numeric_ == NULL) {
+      nlp_->log->printf(hovError, "printf matrix size: %d numeric NULL \n", n_);
+      return -1;
+    }
+    // get sizes
+    const int nnzL = Numeric_->lnz;
+    const int nnzU = Numeric_->unz;
+
+    const int nnzM = (nnzL + nnzU - n_);
+    /* parse the factorization */
+
+    if(mia_ != NULL) {
+      free(mia_);
+    }
+    if(mja_ != NULL) {
+      free(mja_);
+    }
+
+    mia_ = (int*) calloc(sizeof(int), (n_+1));
```

Comment:
Do we really need to reallocate `mia_` and `mja_` on every call? Or could we just `memset(mia_, 0, sizeof(int[n_+1]));` instead?